### ISSUE TITLE
[🐛BUG] GRU4Rec Defaults to Negative Sampling

### ISSUE THREAD

**Issue by jianshen92** [RUCAIBox/RecBole#788]
**Describe the bug**
GRU4Rec Defaults to Negative Sampling with 1 negative sample with `CE` loss function. However, during the training loop, the negative samples are being fed into the model the same as the positive sample. The Interaction `labels` field are not used.

**To Reproduce**
Run `GRU4Rec` model with `ml-100k` dataset.

**Expected behavior**
Expects no negative sample to be fed into GRU4Rec with `CrossEntropy` loss function.

Please correct me if I have any misunderstanding! Thank you!

**Comment by 2017pxy** [comment]
@jianshen92 Hi, if you want to train GRU4Rec without negative sampling, you can set `training_neg_sample_num` to `0` in command line or `.yaml` file.

By the way, if you read the code of GRU4Rec, you will find that when the loss function is `CE`, the negative items will not be used, even though the negative samples are fed into the model.

**Comment by jianshen92** [comment]
> the negative samples will not be used even though the negative samples are fed into the model.
Yes, i stepped through the source code but i could not find the section where negative samples are omitted. Thats why I thought it might be a bug.

Could you point out the section where the negative sample is not used?

**Comment by hyp1231** [comment]
Hi, could you please provide the `.yaml` file you use to run GRU4Rec with ml-100k, we will check it and try to figure out the way to omit negative sampling. Thanks.

**Comment by 2017pxy** [comment]
@jianshen92 Hi, here is the part of GRU4Rec code in [GRU4Rec.py](https://github.com/RUCAIBox/RecBole/blob/master/recbole/model/sequential_recommender/gru4rec.py):
```
def calculate_loss(self, interaction):
        item_seq = interaction[self.ITEM_SEQ]
        item_seq_len = interaction[self.ITEM_SEQ_LEN]
        seq_output = self.forward(item_seq, item_seq_len)
        pos_items = interaction[self.POS_ITEM_ID]
        if self.loss_type == 'BPR':
            neg_items = interaction[self.NEG_ITEM_ID]
            pos_items_emb = self.item_embedding(pos_items)
            neg_items_emb = self.item_embedding(neg_items)
            pos_score = torch.sum(seq_output * pos_items_emb, dim=-1)  # [B]
            neg_score = torch.sum(seq_output * neg_items_emb, dim=-1)  # [B]
            loss = self.loss_fct(pos_score, neg_score)
            return loss
        else:  # self.loss_type = 'CE'
            test_item_emb = self.item_embedding.weight
            logits = torch.matmul(seq_output, test_item_emb.transpose(0, 1))
            loss = self.loss_fct(logits, pos_items)
            return loss
```
You can see that if your loss_type is `BPR`, we will get `neg_items` from `interaction` and use it to calculate the loss. But if the loss_type is `CE`, we will only use `pos_items` to calculate. That's why I said no matter you make negative sampling or not, if your loss_type in GRU4Rec is `CE`, the negative samples will not be used.

**Comment by jianshen92** [comment]
@2017pxy In that case, the validation results for parameter `training_neg_sample_num` = 0 and `training_neg_sample_num` = 1 should be identical. However, running the default settings of with `ml-100k` after 1 epoch, gives mrr 0.0113, when setting neg_sample to 0, i get mrr 0.011.

Anyway I have another question. I couldn't figure exactly why 
at https://github.com/RUCAIBox/RecBole/blob/master/recbole/trainer/trainer.py#L348
you need to do the swapping?
```python
scores[swap_row, swap_col_after] = scores[swap_row, swap_col_before]
```

and at https://github.com/RUCAIBox/RecBole/blob/master/recbole/evaluator/evaluators.py#L68
```python
scores_matrix = torch.flip(scores_matrix, dims=[-1])
```
that you need to flip?


Can we just get the scores matrix at `full_sort_predict` and run `torch.topk` to get the top 10 items?

**Comment by 2017pxy** [comment]
@jianshen92 Hi, about your question, here are my replies.

Firstly, about the result of GRU4Rec:
> In that case, the validation results for parameter training_neg_sample_num = 0 and training_neg_sample_num = 1 should be identical. However, running the default settings of with ml-100k after 1 epoch, gives mrr 0.0113, when setting neg_sample to 0, i get mrr 0.011.

In RecBole,  we will use pytorch random function in both negative sampling and data shuffling (in training phase). Since they all use the pytorch random data list, the random number will be different in data shuffling if you don't make negative sampling. In other words, the random data for data shuffling is different when `training_neg_sample_num` is 1 or 0. That's why the results have a little difference. In my opinion, this kind of change is acceptable, since if you set a different random seed in `.yaml` file, the result will also be different.
 
About the swap and flip of scores:
>Anyway I have another question. I couldn't figure exactly why
at https://github.com/RUCAIBox/RecBole/blob/master/recbole/trainer/trainer.py#L348
you need to do the swapping? and at https://github.com/RUCAIBox/RecBole/blob/master/recbole/evaluator/evaluators.py#L68 that you need to flip?

It's a trick to accelerate the evaluation in RecBole. To get a faster evaluation, we design an evaluation strategy and we will always put positive items' scores at the head of scores matrix. For more details about our evalution design, you can see [here](https://recbole.io/evaluation.html)   

Finally，about your needs:
>Can we just get the scores matrix at full_sort_predict and run torch.topk to get the top 10 items?

We have already implement an API called `case study`, which may be helpful to you. You can read the [case study docs](https://recbole.io/docs/recbole/recbole.utils.case_study.html?highlight=case%20study#) and the [example code](https://github.com/RUCAIBox/RecBole/blob/master/run_example/case_study_example.py) to get more information of this API.

**Comment by jianshen92** [comment]
Thank you @2017pxy for your answering. That clears my understanding!

I would like to also use this thread to ask some extra questions,
are there models in recbole that will use additional interaction features? a.k.a we can load more interaction feature in
```yaml
load_col:
   inter: [user_id, item_id, timestamp, .. additional custom feature]
```

So far the models i have browsed, the `.inter` used are just mainly the item field and rating field.

some example of additional interaction features are:
1. user features that varies across the session user browsing time, time spent on item
2. item feature that varies across the session such as inventory count

**Comment by 2017pxy** [comment]
@jianshen92 Hi, in RecBole, there is a group of models called context-aware recommendation model, which will not only use all kinds of features in `.inter` file, but also use features in `.user` file and `.item` file (if you have). Some famous models like [LR](https://recbole.io/docs/user_guide/model/context/lr.html) and [FM](https://recbole.io/docs/user_guide/model/context/fm.html) belong to this category in RecBole.  If you want to use more features to train a model,  context-aware recommendation models can be a good choice. You can click [here](https://recbole.io/model_list.html) to know about which models are supported in RecBole.

**Comment by rowedenny** [comment]
I am also confused with `loss = CE` and `training_neg_sample_num = 1`. Here is how I am puzzled:

1. when the loss is CE, then the model automatically set `input_type = InputType.POINTWISE`. 
2. As the [neg_sample_minin](https://github.com/RUCAIBox/RecBole/blob/master/recbole/data/dataloader/neg_sample_mixin.py#L107) implemented,  
the corresponding `sampling_func = _neg_sample_by_point_wise_sampling`
3. This `sampling_func` is called  by . [`_neg_sampling `](https://github.com/RUCAIBox/RecBole/blob/df1eefb85607acf1ef747c25ec007a244ca1c090/recbole/data/dataloader/general_dataloader.py#L151). Notice that `self.neg_sample_by` here is 1, thus it will generate negative ids. 
4. When `_neg_sampling ` by `_next_batch_data `, the negative samples are concatenated and used as final batch data.

In short, the batch of data when `loss = CE` and `training_neg_sample_num = 1` will consist of two parts
1. the first half is the positive samples
2. the second half is the same context (here the context means the attributes `item_seq`, etc.), while the only difference is `pos_items`. 

To reproduce this, I print out the `user, item_seq, item_seq_len, pos_items` of interaction in `calculate_loss` of `GRU4Rec` with `train_batch_size=4`. Given the demo, it gives
```
tensor([316, 388, 316, 388], device='cuda:0')
tensor([[ 290, 1226,   53,   50,  178,  410,  428,  407,  615,  358,   90,   69,
          303,  994,   25,  601,   88,  388,  444,  404,   24,   87,   11,  782,
          936,  141,  182,  341,   29,   78,  351,  104,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0],
        [ 747,   79,  717,   85,  802, 1180,  226,  744,  439,   47,   57,  735,
          310,  519,  413,  429,  937,  590,   80,  920,  536,  321, 1277,  287,
          510,   21,  424,  374,  113,  367,   27,  126,  617,   48, 1238,  982,
          103, 1061,   83,  481,  658,  730,  142,  531,  400,  384,  591,  878,
          774,  446],
        [ 290, 1226,   53,   50,  178,  410,  428,  407,  615,  358,   90,   69,
          303,  994,   25,  601,   88,  388,  444,  404,   24,   87,   11,  782,
          936,  141,  182,  341,   29,   78,  351,  104,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0],
        [ 747,   79,  717,   85,  802, 1180,  226,  744,  439,   47,   57,  735,
          310,  519,  413,  429,  937,  590,   80,  920,  536,  321, 1277,  287,
          510,   21,  424,  374,  113,  367,   27,  126,  617,   48, 1238,  982,
          103, 1061,   83,  481,  658,  730,  142,  531,  400,  384,  591,  878,
          774,  446]], device='cuda:0')
tensor([32, 50, 32, 50], device='cuda:0')
tensor([ 578,  785,  399, 1309], device='cuda:0')
```

It is wired since we have half the negative target index as positive, yet the model's performance still looks good. If I misunderstand anything, please be free to let me know.

**Comment by 2017pxy** [comment]
@rowedenny Hi, firstly thanks for your carefully checking and it is indeed a bug. When `loss_type` is `CE` and `training_neg_sample_num` is not 0, the `pos_item` will be mixed with `negative_item`. To be specific, if the `training_neg_sample_num` is 1, the first half of `pos_item` is real positive item while the second half will be negative_items. I must admit that it's a serious problem.

We have already added a check in `configurator.py` (#795) to avoid this bug. However, we are also confused with the model's performance and we will make more tests to explore this question.  

Thanks again for your checking!

**Comment by jianshen92** [comment]
Thanks @rowedenny for the detailed explanation, it was something I wanted to point out but i wasn't being clear enough!
